crates/bls-gadgets/src/y_to_bit.rs

```diff
@@ -83,20 +83,49 @@ impl<P: Bls12Parameters> YToBitGadget<P> {
                 Err(SynthesisError::AssignmentMissing)
             }
         })?;
-        let y_eq_bit = Boolean::alloc(cs.ns(|| "alloc y eq bit"), || {
-            if pk.y.c0.get_value().is_some() {
-                let half = P::Fp::modulus_minus_one_div_two();
-                Ok(pk.y.c0.get_value().get()?.into_repr() == half)
-            } else {
-                Err(SynthesisError::AssignmentMissing)
-            }
-        })?;
+
+        let y_eq_bit =
+            Boolean::alloc(cs.ns(|| "alloc y eq bit"), || {
+                if pk.y.c1.get_value().is_some() {
+                    Ok(pk.y.c1.get_value().get()?.into_repr()
+                        == pk.y.c0.get_value().get()?.into_repr())
+                } else {
+                    Err(SynthesisError::AssignmentMissing)
+                }
+            })?;
+
+        {
+            let neg_c0 = pk.y.c0.negate(cs.ns(|| "neg c0"))?;
+            let lhs = pk.y.c1.add(cs.ns(|| "lhs"), &neg_c0)?;
+            let inv = FpGadget::alloc(cs.ns(|| "alloc (c1 - c0) inv"), || {
+                if pk.y.c1.get_value().is_some() {
+                    Ok(lhs.get_value().get()?.inverse().unwrap_or_else(P::Fp::zero))
+                } else {
+                    Err(SynthesisError::AssignmentMissing)
+                }
+            })?;
+            // (y * y_inv == 1 - y_eq_bit)
+            cs.enforce(
+                || "enforce y_eq_bit",
+                |lc| lhs.get_variable() + lc,
+                |lc| inv.get_variable() + lc,
+                |lc| lc + (P::Fp::one(), CS::one()) + y_eq_bit.lc(CS::one(), P::Fp::one().neg()),
+            );
+            // (y*y_eq_bit == 0)
+            cs.enforce(
+                || "enforce y_eq_bit 2",
+                |lc| pk.y.c1.get_variable() + lc,
```

Comment:
Shouldn't this be `lhs` instead of `y.c1`?

Reply:
Yes, that was the meaning, thanks!
(For what it's worth, lhs == 0 when pky.y.c1 ==0, so that's why the tests still pass).